pkg/storegateway/batch_series_test.go

```diff
@@ -808,6 +811,233 @@ func TestBlockSeriesChunkRefsSetsIterator_ErrorPropagation(t *testing.T) {
+			},
+		},
+		"handles error in addLoad": {
+			existingBlocks: []testBlock{block1, block2},
+			setsToLoad: []seriesChunkRefsSet{
+				{series: []seriesChunkRefs{toSeriesChunkRefs(block1, 0), toSeriesChunkRefs(block1, 1)}},
+				{series: []seriesChunkRefs{toSeriesChunkRefs(block2, 0), toSeriesChunkRefs(block2, 1)}},
+			},
+			expectedSets: []seriesChunksSet{},
+			addLoadErr:   errors.New("test err"),
+			expectedErr:  "test err",
+		},
+		"handles error in load": {
+			existingBlocks: []testBlock{block1, block2},
+			setsToLoad: []seriesChunkRefsSet{
+				{series: []seriesChunkRefs{toSeriesChunkRefs(block1, 0), toSeriesChunkRefs(block1, 1)}},
+				{series: []seriesChunkRefs{toSeriesChunkRefs(block2, 0), toSeriesChunkRefs(block2, 1)}},
+			},
+			expectedSets: []seriesChunksSet{},
+			loadErr:      errors.New("test err"),
+			expectedErr:  "test err",
+		},
+	}
+
+	for testName, testCase := range testCases {
+		testName, testCase := testName, testCase
+		t.Run(testName, func(t *testing.T) {
+			t.Parallel()
+			// Setup
+			bytesPool := &mockedPool{parent: pool.NoopBytes{}}
+			chunkBytes := &pool.BatchBytes{Delegate: bytesPool}
+			readersMap := make(map[ulid.ULID]chunkReader, len(testCase.existingBlocks))
+			for _, block := range testCase.existingBlocks {
+				readersMap[block.ulid] = newFakeChunkReaderWithSeries(block.series, testCase.addLoadErr, testCase.loadErr, chunkBytes)
+			}
+			readers := newChunkReaders(readersMap, chunkBytes, bytesPool)
+
+			// Run test
+			set := newLoadingBatchSet(*readers, newSliceSeriesChunkRefsSetIterator(nil, testCase.setsToLoad...), newSafeQueryStats())
+			loadedSets := readAllSeriesChunksSets(set)
+
+			// Assertions
+			if testCase.expectedErr != "" {
+				assert.ErrorContains(t, set.Err(), testCase.expectedErr)
+			} else {
+				assert.NoError(t, set.Err())
+			}
+			// NoopBytes should allocate slices just the right size, so the packing optimization in BatchBytes should not be used
+			// This allows to assert on the exact number of bytes allocated.
+			var expectedReservedBytes int
+			for _, set := range testCase.expectedSets {
+				for _, s := range set.series {
+					for _, c := range s.chks {
+						expectedReservedBytes += len(c.Raw.Data)
+					}
+				}
+			}
+			assert.Equal(t, expectedReservedBytes, int(bytesPool.balance.Load()))
+
+			// Check that chunks butes are what we expect
+			require.Len(t, loadedSets, len(testCase.expectedSets))
+			for i, loadedSet := range loadedSets {
+				require.Len(t, loadedSet.series, len(testCase.expectedSets[i].series))
+				for j, loadedSeries := range loadedSet.series {
+					testCase.expectedSets[i].series[j].refs = nil
+					assert.Equal(t, testCase.expectedSets[i].series[j], loadedSeries)
+				}
+			}
+
+			// Release the sets and expect that they also return their chunk bytes to the pool
+			for _, s := range loadedSets {
+				s.release()
+			}
+			assert.Zero(t, int(bytesPool.balance.Load()))
+		})
+	}
+}
+
+type fakeChunkReader struct {
```

Comment:
[nit] Instead of "fake", I would call it "mock" which is the naming we're using across the code base.

```suggestion
type chunkReaderMock struct {
```

Reply:
hm.. I learned them with different meanings, so I thought we use whichever fits the purpose better. Will rename here to `mock`